\documentclass{article}
\usepackage[margin=1in]{geometry}
\usepackage{algorithm}
\usepackage{algpseudocode}
\usepackage{amssymb}
\usepackage[backend=biber,style=numeric,sorting=none]{biblatex}
\addbibresource{proposal.bib}
\title{%
	\Large Bees? DNA Motif Discovery with Alternating Global-Local Search  \\
	\large \; \\ - CSC 530: Group 2 Project Proposal -}
\author{Grant Billings and Karthik Sanka}
\date{09/29/2022}
\begin{document}
\maketitle
\section*{\large{Abbreviations}}
\textbf{(l, d)}: a planted motif of length \textit{l} with \textit{d} random changes; \textbf{DNA}: Deoxyribonucleic Acid; \textbf{HMC}: Hamiltonian Monte Carlo; \textbf{MEME}: Multiple Expectation Maximization for Motif Elicitation; \textbf{PSO}: Particle Swarm Optimization;
\section{Executive Summary}
Living organisms have genomes that evolve randomly over time, with natural selection working to increase the frequency of functionally beneficial sequences over generations. Motifs are non-random nucleotide sequences that in many cases have been shown to have biological function in gene regulation. Detection of motifs in sets of sequences is challenging because random mutations make exact matching of sequences ineffective, and brute force methods are very slow. The most popular software package for motif discovery is MEME, which works well but slows down significantly if many query sequences are provided. Motif discovery across large data sets has become an important step in genome analysis. \textit{Software that can efficiently mine these sequences for motifs is needed.}  \\\\
Nature-inspired algorithms have promise for DNA motif discovery since they broadly allow for efficient exploration of potential motifs while allowing good solutions to learn from each other. We propose use of Particle Swarm Optimization with Hamiltonian Monte Carlo (PSO-HMC) in alternating cycles of global and local search to quickly find motifs. Our algorithm will be tuned using implanted motifs in simulated data, tested on previously characterized benchmarking data, and finally applied to discover new sequence motifs in a cotton promoter sequence dataset. Sensitivity, specificity, and running time will be used to compare performance between our software and other widely used alternatives. At the end of the semester, we will present our findings in comparison to other available software in a poster. We will also create an animation showing the algorithm running and share it upload it to Wikipedia so others can gain a visual intuition for how PSO-HMC works. \textit{Our work will contribute to the rapid characterization of large genomic datasets.}
\section{Abstract}
Biologists are interested in detecting motifs from DNA sequencing data because of their role in gene expression and chromatin architecture.
The (l, d) planted motif problem is NP-complete, so heuristics are usually employed to find motifs. Non-probabilistic scoring functions for potential motifs and their positions in sequences are discrete, making the non-convex, non-smooth solution space very difficult to work with using traditional optimization techniques. Nature-inspired algorithms tend to excel in problems of this type due to the ability for the algorithm to exchange information on potential solutions. Here, we propose a novel method for motif discovery using 1) alternating rounds of Particle Swarm Optimization for efficient global exploration of the solution space; and 2) Hamiltonian Monte Carlo for detailed local search to avoid poor outcomes due to local optima. We will implement our algorithm in Julia, and benchmark on synthetic and real datasets. Key deliverables include a poster presentation, as well as release of a graphical representation of the algorithm and code into the public domain. We hope the speed and quality of the predicted motifs will help researchers generate hypotheses for motif sequences that can then be functionally validated through wet lab experiments.
\section{Prior Work}
Example citation \cite{fister2013brief}
\section{Project Description}

\subsection{Data}
\begin{itemize}
	\item Simulated (l,d) implanted motifs
	\item Actual annotated \textit{cis} element plant dataset from Sabastian and Contreras-Moreira, Bioinformatics, Nov 2013
	\item Upstream sequence from fiber-specific genes in Ando et al, BMC Genomics, April 2021
\end{itemize}
\subsection{The Algorithm}
\begin{algorithm}
\caption{Motif Detection with PSO-HMC}
\begin{algorithmic}
\ForAll{$\textrm{motif lengths } k \in 5..15$}
	\State{Initialize a set $M$ of particle position vectors $m$ containing $p$ particles in $\mathbb{Z}^{n}$}
	\While{not converged or $i <$ iteration limit}
		\ForAll{particles $m_{i} \in 1..p$}
			\State{Evaluate the current score for the particle $m_{i}$}
			\State{Sample $j$ particles and their paths near $m_{i}$ with HMC}
			\State{Find the particle $m_{\textrm{best}}^{*}$ in $M$ with the best current score}
			\State{Store $m_{\textrm{best}}^{*}$}
		\EndFor
	\EndWhile
\EndFor
\end{algorithmic}
\end{algorithm}
Sample new particle positions mi+1 from HMC results, with extra probability in the direction of the “best” particle \\
	Report the best z non-overlapping particle positions
\subsection{Implementation}
\begin{itemize}
	\item Implementation in Julia
	\item Particle Swarm Optimization with \textit{inertia} and \textit{social attraction} parameters
	\item Hamiltonian Monte Carlo 
\end{itemize}
\subsection{Experiments}
\subsection{Evaluation and Statistics}
\subsection{Deliverables}
\begin{itemize}
	\item Poster
	\item Graphical demonstration of algorithm
	\item Public availability of code on GitHub
\end{itemize}
\subsection{Anticipated Problems and Solutions}
\section{Timeline}
\printbibliography
\end{document}

